runtime-modules/referendum/src/lib.rs

```diff
@@ -0,0 +1,814 @@
+        let account_id = Self::ensure_regular_user(origin)?;
+
+        let stage = Stage::<T, I>::get();
+
+        // ensure referendum is running
+        let stage_data = match stage {
+            ReferendumStage::Revealing(tmp_stage_data) => (tmp_stage_data),
+            _ => return Err(Error::RevealingNotInProgress),
+        };
+
+        let current_block = <system::Module<T>>::block_number();
+
+        // ensure voting stage is not expired (it can happend when superuser haven't call `finish_voting_start_revealing` yet)
+        if current_block >= T::RevealStageDuration::get() + stage_data.start + One::one() {
+            return Err(Error::RevealingNotInProgress);
+        }
+
+        // ensure account haven't voted yet
+        if !Votes::<T, I>::contains_key(&account_id) {
+            return Err(Error::NoVoteToReveal);
+        }
+
+        if vote_option_index >= &stage_data.options_count {
+            return Err(Error::InvalidVote);
+        }
+
+        let sealed_vote = Votes::<T, I>::get(&account_id);
+
+        // ensure vote was casted for the running referendum
+        if cycle_id != sealed_vote.cycle_id {
+            return Err(Error::InvalidVote);
+        }
+
+        // ensure commitment corresponds to salt and vote option
+        let commitment = R::calculate_commitment(&account_id, salt, &cycle_id, vote_option_index);
+        if commitment != sealed_vote.commitment {
+            return Err(Error::InvalidReveal);
+        }
+
+        Ok((account_id, sealed_vote))
+    }
+
+    fn can_release_stake(origin: T::Origin) -> Result<T::AccountId, Error<T, I>> {
+        fn voted_for_winner_last_cycle<T: Trait<I>, I: Instance>(
+            last_cycle_result: ReferendumResult<u64, T::VotePower>,
+            option_voted_for: Option<u64>,
+        ) -> bool {
+            let voted_for = match option_voted_for {
+                Some(tmp_vote) => tmp_vote,
+                None => return false,
+            };
+
+            let previous_winners = match last_cycle_result {
+                ReferendumResult::Winners(tmp_winners) => tmp_winners,
+                ReferendumResult::ExtraWinners(tmp_winners) => tmp_winners,
+                ReferendumResult::NotEnoughWinners(tmp_winners) => tmp_winners,
+                ReferendumResult::NoVotesRevealed => vec![],
+            };
+
+            let voted_for_winner = previous_winners.iter().position(|item| item.0 == voted_for);
+
+            voted_for_winner.is_some()
+        }
+
+        let cycle_id = CurrentCycle::<I>::get();
+
+        // ensure superuser requested action
+        let account_id = Self::ensure_regular_user(origin)?;
+
+        // ensure there is some vote with locked stake
+        if !Votes::<T, I>::contains_key(&account_id) {
+            return Err(Error::NoStakeLocked);
+        }
+
+        let sealed_vote = Votes::<T, I>::get(&account_id);
+
+        // enable stake release only during
+        if cycle_id == sealed_vote.cycle_id {
```

Comment:
This appears to allow someone to release stake if they voted in this cycle and its during voting.

Some of this logic will likely be lifted out, but the rule for the council use case is
- if its a future cycle, its always ok.
- if its the same cycle, then we are in the idle period, and you voted for the loser.

Reply:
Those two condition will be satisfied in the council.

This particular condition is tied with the following `match` on the next line - it enables user to revoke his vote during voting period.